### ISSUE TITLE
MXNET can't exit normally

### ISSUE THREAD

**Issue by dytmas** [apache/incubator-mxnet#5511]
CUDA version: 375.39
cudnn: 5.1.10
MXNET version: newest by now, installed through source code.
gpu: nvidia 1080

code:
  22 net = mx.symbol.Variable(name='data')
  23 net = mx.symbol.FullyConnected(data=net, name='fc1', num_hidden=10)
  24 net = mx.symbol.SoftmaxOutput(data=net, name='softmax')
  25
  26 model = mx.mod.Module(net, context=mx.gpu(0))
  27
  28
  29 #print model.label_names
  30 training_X = np.random.rand(num_batches*batch_size, 1000)
  31 training_Y = np.random.randint(0, 10, size=(num_batches*batch_size,))
  32
  33
  34 print('data_generated')
  35 train_iter = mx.io.NDArrayIter(training_X, \
  36         label=training_Y, batch_size=batch_size)
  37
  38 print('iterator allocated!')
  39
  40 #exit()
  41
  42 #print 'before'
  43 #testX, testY = next(train_iter)
  44 #print 'after'
  45 #print testX, testY
  46
  47 last_time = time.time()
  48
  49 def batch_callback(batch_obj):
  50     if batch_obj.nbatch % 100 == 0 and batch_obj.nbatch != 0:
  51         global last_time
  52         print('step %d: %f images/s' % (batch_obj.nbatch, 100*batch_size / (time.time()-last_time)))
  53         last_time = time.time()
  54
  55 model.fit(
  56         train_data         = train_iter,
  57         optimizer          = 'sgd',
  58         optimizer_params   = {'learning_rate':learning_rate},
  59         eval_metric = [],
  60         num_epoch          = 10,  #num_epochs,
  61         batch_end_callback = batch_callback)

Sometimes, the code exits normally, sometimes, it got crashed after training:

[02:07:24] /home/xxx/mxnet/dmlc-core/include/dmlc/./logging.h:300: [02:07:24] /home/xxx/mxnet/mshadow/mshadow/./stream_gpu-inl.h:125: Check failed: err == CUDNN_STATUS_SUCCESS (4 vs. 0) CUDNN_STATUS_INTERNAL_ERROR

Stack trace returned 6 entries:
[bt] (0) /home/xxx/anaconda2/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet/libmxnet.so(_ZN4dmlc15LogMessageFatalD1Ev+0x29) [0x7fb8cc13e369]
[bt] (1) /home/xxx/anaconda2/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet/libmxnet.so(_ZN7mshadow9NewStreamINS_3gpuEEEPNS_6StreamIT_EEbb+0x135) [0x7fb8ccc41fe5]
[bt] (2) /home/xxx/anaconda2/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet/libmxnet.so(_ZNSt17_Function_handlerIFvvEZZN5mxnet6engine23ThreadedEnginePerDevice13PushToExecuteEPNS2_8OprBlockEbENKUlvE1_clEvEUlvE_E9_M_invokeERKSt9_Any_data+0x87) [0x7fb8ccc50ad7]
[bt] (3) /home/xxx/anaconda2/bin/../lib/libstdc++.so.6(+0xb4870) [0x7fb8be7ad870]
[bt] (4) /lib64/libpthread.so.0(+0x7dc5) [0x7fb8e48eedc5]
[bt] (5) /lib64/libc.so.6(clone+0x6d) [0x7fb8e3f1473d]

[02:07:24] /home/xxx/mxnet/dmlc-core/include/dmlc/./logging.h:300: [02:07:24] /home/xxx/mxnet/mshadow/mshadow/./stream_gpu-inl.h:125: Check failed: err == CUDNN_STATUS_SUCCESS (4 vs. 0) CUDNN_STATUS_INTERNAL_ERROR

Stack trace returned 6 entries:
[bt] (0) /home/xxx/anaconda2/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet/libmxnet.so(_ZN4dmlc15LogMessageFatalD1Ev+0x29) [0x7fb8cc13e369]
[bt] (1) /home/xxx/anaconda2/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet/libmxnet.so(_ZN7mshadow9NewStreamINS_3gpuEEEPNS_6StreamIT_EEbb+0x135) [0x7fb8ccc41fe5]
[bt] (2) /home/xxx/anaconda2/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet/libmxnet.so(_ZNSt17_Function_handlerIFvvEZZN5mxnet6engine23ThreadedEnginePerDevice13PushToExecuteEPNS2_8OprBlockEbENKUlvE1_clEvEUlvE_E9_M_invokeERKSt9_Any_data+0x87) [0x7fb8ccc50ad7]
[bt] (3) /home/xxx/anaconda2/bin/../lib/libstdc++.so.6(+0xb4870) [0x7fb8be7ad870]
[bt] (4) /lib64/libpthread.so.0(+0x7dc5) [0x7fb8e48eedc5]
[bt] (5) /lib64/libc.so.6(clone+0x6d) [0x7fb8e3f1473d]

terminate called after throwing an instance of 'dmlc::Error'
  what():  [02:07:24] /home/xxx/mxnet/mshadow/mshadow/./stream_gpu-inl.h:125: Check failed: err == CUDNN_STATUS_SUCCESS (4 vs. 0) CUDNN_STATUS_INTERNAL_ERROR

Stack trace returned 6 entries:
[bt] (0) /home/xxx/anaconda2/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet/libmxnet.so(_ZN4dmlc15LogMessageFatalD1Ev+0x29) [0x7fb8cc13e369]
[bt] (1) /home/xxx/anaconda2/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet/libmxnet.so(_ZN7mshadow9NewStreamINS_3gpuEEEPNS_6StreamIT_EEbb+0x135) [0x7fb8ccc41fe5]
[bt] (2) /home/xxx/anaconda2/lib/python2.7/site-packages/mxnet-0.9.4-py2.7.egg/mxnet/libmxnet.so(_ZNSt17_Function_handlerIFvvEZZN5mxnet6engine23ThreadedEnginePerDevice13PushToExecuteEPNS2_8OprBlockEbENKUlvE1_clEvEUlvE_E9_M_invokeERKSt9_Any_data+0x87) [0x7fb8ccc50ad7]
[bt] (3) /home/xxx/anaconda2/bin/../lib/libstdc++.so.6(+0xb4870) [0x7fb8be7ad870]
[bt] (4) /lib64/libpthread.so.0(+0x7dc5) [0x7fb8e48eedc5]
[bt] (5) /lib64/libc.so.6(clone+0x6d) [0x7fb8e3f1473d]

terminate called recursively

=============================================================
sometimes, it just can't exit after training (the memory can't be released).

**Comment by dytmas** [comment]
Same program, 3 different outcomes!

**Comment by dytmas** [comment]
@mli @tqchen it's a bug of mxnet or other libraries or the driver?

**Comment by xwzheng1020** [comment]
I met the similar problem on windows platform, no gpu used.
After running example\image-classification\predict-cpp\image-classification-predict.cc, the program cannot exit by itself.

**Comment by cjolivier01** [comment]
@xwzheng1020 this happens consistently? What's the command-line?

**Comment by cjolivier01** [comment]
Fix:  https://github.com/dmlc/mxnet/pull/6773

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
